napari/components/_dims/model.py

```diff
@@ -1,143 +1,297 @@
-        self._max_shape = _compute_max_shape(shapes, self.max_dims)
+        if self.display[axis] != display:
+            self.display[axis] = display
+            self.events.axis(axis=axis)
 
-    def _on_layers_change(self, event):
-        """Called whenever a layer is changed.
+    def _set_2d_viewing(self):
+        """Sets the 2d viewing
         """
-        self._child_layer_changed = True
-        self._update()
+        self.display = [False] * len(self.display)
+        self.display[-1] = True
+        self.display[-2] = True
+
+    def _add_axes(self, axis: int, no_event=None):
+        """Makes sure that the given axis is in the dimension model
+
+        Parameters
+        ----------
+        axis : int
+            Dimension index
 
-    def _update(self):
-        """Updates the viewer.
+        Returns
+        -------
+        dimensions : list
+            List of axes
         """
-        if self._child_layer_changed:
-            self._child_layer_changed = False
-            self._recalc_max_dims = True
-            self._recalc_max_shape = True
-            self._need_slider_update = True
-
-        if self._need_redraw:
-            self._need_redraw = False
-            self.viewer._update_layers()
-
-        if self._recalc_max_dims:
-            self._recalc_max_dims = False
-            self._calc_max_dims()
-
-        if self._recalc_max_shape:
-            self._recalc_max_shape = False
-            self._calc_max_shape()
-
-        if self._need_slider_update:
-            self._need_slider_update = False
-            self._update_sliders()
+        if axis >= self.ndims:
+            old_ndims = self.ndims
+            margin_length = 1 + axis - self.ndims
+            self.range.extend([(0.0, 1.0, 0.01)] * (margin_length))
+            self.point.extend([0.0] * (margin_length))
+            self.interval.extend([(0.3, 0.7)] * (margin_length))
+            self.mode.extend([DimsMode.POINT] * (margin_length))
+            self.display.extend([False] * (margin_length))
+
+            if not no_event:
+                # Notify listeners that the number of dimensions have changed
+                self.events.ndims()
+
+                # Notify listeners of which dimensions have been affected
+                for axis_changed in range(old_ndims - 1, self.ndims):
+                    self.events.axis(axis=axis_changed)
+
+            return list(range(old_ndims, 1 + axis))
+
+        return []
+
+    def _trim_ndims(self, ndims: int):
+        """This internal method is used to trim the number of axis.
+
+        Parameters
+        ----------
+        ndims : int
+            The new number of dimensions
+        """
+        if ndims < self.ndims:
+            self.range = self.range[:ndims]
+            self.point = self.point[:ndims]
+            self.interval = self.interval[:ndims]
+            self.mode = self.mode[:ndims]
+            self.display = self.display[:ndims]
```

Comment:
Are we always trimming from the back? I actually think it's more likely that we are trimming from the front... If we display a 2D max projection image on top of a 3D slice, the natural max projection is along axis 0 and so the remaining axes will be right-aligned. This is also how NumPy broadcasting is supposed to work (right alignment).

Reply:
i am switching to trim from the front, but it's also unclear how / why this method will ever get used ....

Reply:
I you remove a layer that has more dimensions that all remaining layers... 
For now we don't really support having layers with a different number of
dimensions, but eventually, when we support it, this could come handy.
In general, it is good design to have models be robust and complete so 
they support not just 'adding'/'incrementing' but also 'removing'/'decrementing.

Reply:
fair point. I think trimming from the front is better though, especially in that case. I have moved all of this onto the `ndim` setter too. Let me know if that is not ok